### PR TITLE
A0-1236: Minting and burning reward tokens

### PR DESCRIPTION
**Pull request by h4nsu**
# Description

Turns on Mintable and Burnable extensions of PSP22 standard for game_token contracts. Reward tokens are now minted directly to user's account when `press` is called and burned from their account when spent on the marketplace.

## Type of change

- New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Review by pmikolajczyk41 — APPROVED** [review]
image

File: contracts/game_token/lib.rs [hunk 1]
```diff
@@ -14,19 +14,22 @@ pub mod game_token {
         codegen::{EmitEvent, Env},
         reflect::ContractEventBase,
     };
-    use ink_prelude::{format, string::String};
+    use ink_prelude::{format, string::String as Strink};
```
[the remark this answers is not in the document]
Comment (fbielejec) [reply contracts/game_token/lib.rs:17]: Stink?

File: contracts/game_token/lib.rs [hunk 2]
```diff
@@ -14,19 +14,22 @@ pub mod game_token {
...
     use ink_storage::traits::SpreadAllocate;
     use openbrush::{
-        contracts::psp22::{extensions::metadata::*, Internal},
-        traits::Storage,
+        contracts::psp22::{
+            extensions::{burnable::*, metadata::*, mintable::*},
+            Internal,
+        },
+        traits::{Storage, String},
```
Comment (fbielejec) [inline contracts/game_token/lib.rs:24]: a wot szto, traits::String?

Comment (pmikolajczyk41) [reply contracts/game_token/lib.rs:24]: it's alias for `Vec<u8>` xD

File: contracts/game_token/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (h4nsu) [reply contracts/game_token/lib.rs:17]: string+ink = strink :P
As long as we have "cliain" I'm not changing that XD

File: contracts/game_token/lib.rs (hunk 2 above)
Comment (h4nsu) [reply contracts/game_token/lib.rs:24]: Conveniently defined in `traits` module, just to, you know, be easier to follow :P

File: contracts/access_control/roles.rs [hunk 3]
```diff
@@ -10,10 +10,14 @@ use scale::{Decode, Encode};
...
-    /// Indicates account can initialize a contract from a given code hash.
+    /// Indicates account that can initialize a contract from a given code hash.
     Initializer(Hash),
-    /// Indicates account can add liquidity to a DEX contract (call certain functions)
+    /// Indicates account that can add liquidity to a DEX contract (call certain functions)
     LiquidityProvider(AccountId),
+    /// Indicates account that can mint tokens of a given token contract,
+    Minter(AccountId),
+    /// Indicates account that can burn tokens of a given token contract,
+    Burner(AccountId),
```
[the remark this answers is not in the document]
Comment (h4nsu) [reply contracts/access_control/roles.rs:22]: My current understanding is none :P It probably boils down to how enums are represented internally and encoded/decoded

**Review by fbielejec — COMMENTED** [review]
Works flawlessly, my only comment is to settle for one bloody String if we can, and fix the CI failing (c-c version update will do)

File: contracts/access_control/roles.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (fbielejec) [reply contracts/access_control/roles.rs:22]: our contracts check for enum variant - so as long as it's not removed it's golden. If someone is switching on variants, and does not have a catch-all clause - might be a problem.

File: contracts/scripts/clean.sh [hunk 4]
```diff
@@ -10,7 +10,7 @@ function terminate_contract {
   local contract_address=$(get_address $contract_name)
 
   cd "$CONTRACTS_PATH"/"$contract_dir"
-  cargo contract call --url "$NODE" --contract $contract_address --message terminate --suri "$AUTHORITY_SEED"
+  cargo contract call --url "$NODE" --contract $contract_address --message terminate --suri "$AUTHORITY_SEED" --skip-confirm
```
[the remark this answers is not in the document]
Comment (h4nsu) [reply contracts/scripts/clean.sh:13]: done

File: contracts/scripts/deploy.sh [hunk 5]
```diff
@@ -74,9 +75,8 @@ function deploy_game_token {
 
   cd "$CONTRACTS_PATH"/game_token
 
-  # TODO : remove balance when token is mintable
```
[the remark this answers is not in the document]
Comment (h4nsu) [reply contracts/scripts/deploy.sh:77]: Personally, I'm all for TODOs in code. They are super useful

**Review by fbielejec — APPROVED** [review]
good to go!

**Pull request merged**